File: python/day02/day2.py
import pathlib
import re

CWD = pathlib.Path(__file__).parent.absolute()
filename = pathlib.PurePath(CWD, 'data')
with open(filename) as f:
    count1 = count2 = 0
    data = [x for x in f.readlines()]
    for x in data:
        low, high, letter, word = re.search(r'(\d+)-(\d+) (\w): (\w+)', x).groups()
        low = int(low)
        high = int(high)

        if low <= word.count(letter) <= high:
            count1 += 1

        if (word[low-1], word[high-1]).count(letter) == 1:
            count2 += 1


print(f'part 1: {count1}')
print(f'part 2: {count2}')

File: js/day02/day2.js
import fs from 'fs'
import path from 'path'

let filename = path.join(path.dirname(process.argv[1]), 'data')

let data = new Set(
    fs.readFileSync(filename, 'utf8')
        .split('\n')
        .filter(x => x !== '')
        .map(item => {
            let reg = /(\d+)-(\d+) (\w): (\w+)/
            let match = reg.exec(item)
            let [_, low, high, letter, word] = match
            return [low, high, letter, word]
        })
)

let count1 = 0
let count2 = 0

for (let [low, high, letter, word] of data) {
    low = parseInt(low)
    high = parseInt(high)

    let exp = new RegExp(`${letter}`, 'g')
    let lcount = (word.match(exp) || []).length
    if (low <= lcount && lcount <= high) {
        count1 += 1
    }

    if ((word[low-1] + word[high-1]).includes(letter) && word[low-1] !== word[high-1]) {
        count2 += 1
    }
}

console.log(count1)
console.log(count2)
